File: Backend/app/models/product.py
from typing import List, Optional, Union
from bson import ObjectId
from pydantic import BaseModel, Field, field_validator
from pydantic_core.core_schema import ValidationInfo
from pydantic import GetCoreSchemaHandler

# --- Giữ nguyên PyObjectId của bạn ---
class PyObjectId(ObjectId):
    @classmethod
    def __get_pydantic_core_schema__(cls, source, handler: GetCoreSchemaHandler):
        def validate(value, info: ValidationInfo):
            if not ObjectId.is_valid(value):
                raise ValueError("Invalid ObjectId")
            return ObjectId(value)
        
        return handler(source).copy(update={"function": validate})

class Product(BaseModel):
    id: Optional[str] = Field(alias="_id", default=None)
    name: str
    description: Optional[str] = ""
    category: Union[List[str], str, None] = None 
    rates: Optional[float] = None
    price: Optional[float] = 0.0
    distance: Optional[Union[float, str]] = None 
    calories: Optional[Union[int, str]] = None  
    preparationTime: Optional[str] = None
    tags: Optional[List[str]] = [] 
    imageURL: Optional[str] = None

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}
        populate_by_name = True

    @field_validator('category', mode='before')
    @classmethod
    def parse_category(cls, v):
        if isinstance(v, str):
            return [v] 
        return v

File: ecommercial_shopping/lib/core/models/product.dart
class Product {
  final String id;
  final String name;
  final String description;
  final double rates;
  final double price;
  final double distance;
  final int calories;
  final String preparationTime;
  final List<String> tags;
  final String imageUrl;
  final List<String> category;

  Product({
    required this.id,
    required this.name,
    required this.description,
    required this.rates,
    required this.price,
    required this.distance,
    required this.calories,
    required this.preparationTime,
    required this.tags,
    required this.imageUrl,
    required this.category,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      rates: (json['rates'] as num?)?.toDouble() ?? 0.0,
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      distance: (json['distance'] as num?)?.toDouble() ?? 0.0,
      calories: json['calories'] ?? 0,
      preparationTime: json['preparationTime'] ?? '',
      tags: List<String>.from(json['tags'] ?? []),
      imageUrl: json['imageURL'] ?? '',
      category: List<String>.from(json['category'] ?? []),
    );
  }
}
